### PR TITLE
Fix RR policy connectivity state upon subchannels shutdown

### LINKED ISSUE #11578: No reconnect attempts with round_robin
It appears that when using the `round_robin` client-side load-balancing policy, once a "subchannel" (i.e. one of the server instances) is down, the client never attempts to reconnect again.

This is easy to reproduce even in the trivial case of a single client and server, where the client makes repeated requests through a channel configured with `round_robin` that specifies only the single server instance. When bringing the server down and back up, the client is never again able to make a request to it.

Contrast this with the default ("pick first") load balancing policy: here, the client successfully reconnects.

I have tried reducing the `grpc.min_reconnect_backoff_ms` and `grpc.initial_reconnect_backoff_ms` config values, but to no avail.

I'm using gRPC 1.4.0. (In 1.3.1, this seemed to work at least partially, with the client periodically attempting to reconnect, although I often observed it not actually succeeding due to some spurious-seeming timeout error.)

Am I correct in thinking this is a bug? Or is it intentional that no reconnect attempts are ever made?

### PR DESCRIPTION
**Pull request by dgquintas**
Fixes #11578

### CODE REVIEW

File: src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c [hunk 1]
```diff
@@ -425,14 +427,20 @@ static void update_state_counters_locked(subchannel_data *sd) {
...
     GPR_ASSERT(subchannel_list->num_transient_failures > 0);
     --subchannel_list->num_transient_failures;
+  } else if (sd->prev_connectivity_state == GRPC_CHANNEL_SHUTDOWN) {
+    GPR_ASSERT(subchannel_list->num_shutdown > 0);
+    --subchannel_list->num_shutdown;
   } else if (sd->prev_connectivity_state == GRPC_CHANNEL_IDLE) {
     GPR_ASSERT(subchannel_list->num_idle > 0);
     --subchannel_list->num_idle;
   }
+
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c]: Please avoid blank lines within functions.

**Comment by dgquintas** [comment]
@markdroth, PTAL. The improvements to the fake resolver surfaced a bug in pick first. The new Fake Resolver returns the last available resolution in "saw error" when PF goes into SHUTDOWN as part of the `UpdatesTest.UpdateBalancersDeadUpdate` fixture, that kills all balancers PF cares about. Previously, the fake resolver wouldn't re-resolve upon "saw error"/PF-shutdown. When it does, the re-resolution from "saw error" happens shortly before the new resolution from the test, properly exercising the "concurrent" updates. Consider:

```
[...]
I0627 14:42:28.329300483    1042 grpclb_end2end_test.cc:824] ********** KILLED BALANCER 0 *************
I0627 14:42:28.329308293    1042 grpclb_end2end_test.cc:827] ========= BEFORE SECOND BATCH ==========
D0627 14:42:28.329683613    1042 pick_first.c:468]           Pick First 0x611000004b40 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (0 total); State: 3; 

[first PF instance going into shutdown because we killed the balancer]

D0627 14:42:28.329739764    1042 pick_first.c:686]           Pick First 0x611000010bc0 created.
I0627 14:42:28.329750394    1042 pick_first.c:334]           Pick First 0x611000010bc0 received update with 3 addresses
I0627 14:42:28.329883706    1042 pick_first.c:415]           Pick First 0x611000010bc0 created subchannel 0x616000003f80 for address uri ipv4:127.0.0.1:30706
I0627 14:42:28.329994559    1042 pick_first.c:415]           Pick First 0x611000010bc0 created subchannel 0x616000003c80 for address uri ipv4:127.0.0.1:17008
I0627 14:42:28.330073121    1042 pick_first.c:415]           Pick First 0x611000010bc0 created subchannel 0x616000003980 for address uri ipv4:127.0.0.1:7161

[client channel creates a new PF instance. Note that this new instance is using the re-resolution from "saw error". Previously, no PF would created yet because "saw error" was effectively a no-op. Next line is the ultimate destruction of the previous instance (cb delay)]

D0627 14:42:28.330120534    1042 pick_first.c:99]            Pick First 0x611000004b40 destroyed.

I0627 14:42:28.334477135    1042 grpclb_end2end_test.cc:829] ========= DONE WITH SECOND BATCH ==========
I0627 14:42:28.334515874    1042 grpclb_end2end_test.cc:852] ========= ABOUT TO UPDATE 1 ==========
I0627 14:42:28.334675493    1042 pick_first.c:334]           Pick First 0x611000010bc0 received update with 1 addresses

[this is the update from the test]

I0627 14:42:28.334804465    1042 pick_first.c:415]           Pick First 0x611000010bc0 created subchannel 0x616000001b80 for address uri ipv4:127.0.0.1:17008
D0627 14:42:28.334820287    1042 pick_first.c:275]           Pick First 0x611000010bc0 unsubscribing from subchannel 0x616000003f80

[because we got an update, we are discarding the previous, still-no-fully-finalized (not "selected") update, the one from saw error]

D0627 14:42:28.335021268    1042 pick_first.c:468]           Pick First 0x611000010bc0 connectivity changed. Updating selected: 0; Updating subchannels: 1; Checking 0 index (3 total); State: 1; 
I0627 14:42:28.335039913    1042 grpclb_end2end_test.cc:854] ========= UPDATE 1 DONE ==========
I0627 14:42:28.335336392    1042 subchannel.c:682]           Connect failed: {"created":"@1498599748.335192710","description":"Failed to connect to remote host: OS Error","errno":111,"file":"src/core/lib/iomgr/tcp_client_posix.c","file_line":200,"os_error":"Connection refused","syscall":"connect","target_address":"ipv4:127.0.0.1:30706"}
D0627 14:42:28.335364295    1042 pick_first.c:468]           Pick First 0x611000010bc0 connectivity changed. Updating selected: 0; Updating subchannels: 1; Checking 0 index (3 total); State: 3; 

[the unsubscription CB is invoked with an error set not to CANCELLED, but to the actual error encountered while trying to connect to the first "saw error" update, which fails because the test killed the balancer]

D0627 14:42:28.335429691    1042 pick_first.c:499]           Pick First 0x611000010bc0 unreffing subchannel 0x616000003f80
D0627 14:42:28.335502862    1042 pick_first.c:499]           Pick First 0x611000010bc0 unreffing 
 subchannel 0x616000003c80
D0627 14:42:28.335559008    1042 pick_first.c:499]           Pick First 0x611000010bc0 unreffing 
 subchannel 0x616000003980

[death throes of the "saw error" update]

D0627 14:42:28.335738537    1042 pick_first.c:468]           Pick First 0x611000010bc0 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (1 total); State: 1; 
D0627 14:42:28.336079129    1042 pick_first.c:468]           Pick First 0x611000010bc0 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (1 total); State: 2; 
I0627 14:42:28.336094510    1042 pick_first.c:572]           Pick First 0x611000010bc0 selected subchannel 0x616000001b80 (connected 0x612000005140)

[the actual, latest update, from the test]

I0627 14:42:28.336121918    1042 pick_first.c:585]           Servicing pending pick with selected subchannel 0x612000005140
```

Reiterating, the bug in PF was the assumption that the callback corresponding to the unsubscription from subchannel connectivity updates is always invoked with CANCELLED. It will in absence of another error. In this case, the error exists: the failure to connect to a defunct load balancer from the "saw error" re-resolution the fake resolver now performs.

**Comment by dgquintas** [comment]
Word.

**Comment by dgquintas** [comment]
Issues: #11223 #11654 #11522

**Pull request merged**